### PR TITLE
[request-promise]: Adding generics to request promise API

### PR DESCRIPTION
**Pull request by puneetar**
Please fill in this template.

- [ ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Review by vansergen — COMMENTED** [review]
@puneetar Could you give an example? Why do you want to bring these changes? It seems like a BREAKING change. Because now I can do

```typescript
import { RequestPromiseAPI } from "request-promise";
export type A = RequestPromiseAPI;
```

But with your changes I will have to specify the type:
```typescript
import { RequestPromiseAPI } from "request-promise";
export type A = RequestPromiseAPI<any>;
```

File: types/request-promise/index.d.ts [hunk 1]
```diff
@@ -29,11 +29,11 @@ declare namespace requestPromise {
         resolveWithFullResponse?: boolean;
     }
 
-    type RequestPromiseAPI = request.RequestAPI<RequestPromise, RequestPromiseOptions, request.RequiredUriUrl>;
+    type RequestPromiseAPI<T> = request.RequestAPI<RequestPromise<T>, RequestPromiseOptions, request.RequiredUriUrl>;
```
Comment (vansergen) [inline types/request-promise/index.d.ts]: Let's change it to this:
```typescript
type RequestPromiseAPI<T = any> = request.RequestAPI<RequestPromise<T>, RequestPromiseOptions, request.RequiredUriUrl>;
```

File: types/request-promise/index.d.ts [hunk 2]
```diff
@@ -29,11 +29,11 @@ declare namespace requestPromise {
...
     type OptionsWithUri = request.UriOptions & RequestPromiseOptions;
     type OptionsWithUrl = request.UrlOptions & RequestPromiseOptions;
     type Options = OptionsWithUri | OptionsWithUrl;
 }
 
-declare const requestPromise: requestPromise.RequestPromiseAPI;
+declare const requestPromise: requestPromise.RequestPromiseAPI<any>;
```
Comment (vansergen) [inline types/request-promise/index.d.ts]: We do not need to specify `any` here

**Review by vansergen — CHANGES_REQUESTED** [review]

**Comment by puneetar** [comment]
@vansergen  Thanks for the suggestion, that is a better approach. I updated the PR.

**Review by vansergen — APPROVED** [review]
LGTM

**Review by orta — APPROVED** [review]

**Pull request merged**